User secrets feature

Praburaj opened the pull request:
@loudej @DamianEdwards @muratg

Praburaj commented:
FYI @davidfowl

Praburaj commented:
@rustd

anurse commented on src/Microsoft.Framework.ConfigurationModel.UserSecrets/PathHelper.cs:
```diff
@@ -0,0 +1,50 @@
+{
+    public class PathHelper
+    {
+        private const string SECRETS_FILE_NAME = "secrets.json";
```
We generally use "TitleCase" for constants.

Praburaj replied:
sure. I'll fix it.

Praburaj commented on src/Microsoft.Framework.ConfigurationModel.UserSecrets/PathHelper.cs:
```diff
@@ -0,0 +1,50 @@
+            var root = Environment.GetEnvironmentVariable("APPDATA") ??         // On Windows it goes to %APPDATA%\Microsoft\UserSecrets\
+                        Environment.GetEnvironmentVariable("HOME");             // On Mac/Linux it goes to ~/Microsoft/UserSecrets/
+
+            return Path.Combine(root, "Microsoft", "UserSecrets", userSecretsId, SECRETS_FILE_NAME);
```
I think I can lower case `Microsoft` and `UserSecrets`.

anurse replied:
On Mac and Linux "Microsoft/UserSecrets" should probably be ".microsoft/usersecrets". Generally, folders in `HOME` that are for specific applications/frameworks/etc. are in folder names prefixed with `.` to keep the non-`.` folders for users. Also, *nix file systems generally have lowercase file paths.

anurse replied:
Ah, you already commented something basically the same :). If possible, it should be "Microsoft" and "UserSecrets" on Windows and ".microsoft" and "usersecrets" on Mac/Linux, to fit in with other things on those platforms.

Praburaj replied:
Hmm. To keep it simple I'll have to assume the current OS is `Windows` if environment variable `APPDATA` is available and assume it is non-windows on its absense. We need a better way to figure out the OS. I'll file a work item in Configuration repo to update it when we have the `OSVersion` available across.

anurse commented:
Looks good to me, others should probably give you the ship-it squirrel as I'm not 100% up on it.

ChengTian commented on src/SecretManager/Program.cs:
```diff
@@ -0,0 +1,212 @@
+        {
+            try
+            {
+                var app = new CommandLineApplication();
```
```c#
app.Name = "user-secret"
```

Praburaj replied:
Fixed this.

ChengTian commented on src/SecretManager/Program.cs:
```diff
@@ -0,0 +1,212 @@
+
+                app.Command("set", c =>
+                {
+                    var optionProject = c.Option("-p|--project <PATH>", "Path to the project", CommandOptionType.SingleValue);
```
use `c.Description = "..."` to add description for each command

ChengTian commented on src/SecretManager/Program.cs:
```diff
@@ -0,0 +1,212 @@
+                app.Command("set", c =>
+                {
+                    var optionProject = c.Option("-p|--project <PATH>", "Path to the project", CommandOptionType.SingleValue);
+                    var keyArg = c.Argument("<name>", "Name of the key");
```
I would suggest we follow the convention used by other command-line tools like `dnu`, use `[name]` to represent argument value name

ChengTian replied:
Same comments for all arguments below

Praburaj commented:
Good points @ChengTian. Fixed all of those and updated the PR. Command line help looks like below. 

```
Usage: user-secret [options] [command]

Options:
  -?|-h|--help  Show help information
  -v|--verbose  Verbose output

Commands:
  set     Sets the user secret to the specified value
  help    Show help information
  remove  Removes the specified user secret
  list    Lists all the application secrets
  clear   Deletes all the application secrets

Use "user-secret help [command]" for more information about a command.
```

Help for a sub-command looks like this:
```
C:\Users\username>user-secret clear --help
Usage: user-secret clear [options]

Options:
  -p|--project <PATH>  Path to project, default is current directory
  -?|-h|--help         Show help information

```

glennc commented on src/Microsoft.Framework.ConfigurationModel.UserSecrets/ConfigurationExtensions.cs:
```diff
@@ -0,0 +1,33 @@
+        /// <returns></returns>
+        public static IConfigurationSourceRoot AddUserSecrets([NotNull]this IConfigurationSourceRoot configuration)
+        {
+            var appEnv = (IApplicationEnvironment)CallContextServiceLocator.Locator.ServiceProvider.GetService(typeof(IApplicationEnvironment));
```
Make sure we add a note in https://github.com/aspnet/Configuration/pull/160 so that this doesn't slip through.

glennc commented on src/Microsoft.Framework.ConfigurationModel.UserSecrets/PathHelper.cs:
```diff
@@ -0,0 +1,57 @@
+
+        public static string GetSecretsPath([NotNull]string projectPath)
+        {
+            var projectFilePath = Path.Combine(projectPath, "project.json");
```
@davidfowl @loudej We talked once about providing a property bag on IApplicationEnvironment that contains all the properties in the project.json, so that things like this could flow. If we did that then we wouldn't have to parse the project.json again here. I like that better than parsing this again. What do you think?

davidfowl replied:
Sure but that's not going to happen right now.

glennc replied:
Sure. But I will make you an issue for it ;)

Praburaj replied:
Yeah hosting also reads webroot from project json directly. It will be helpful.

glennc commented on src/Microsoft.Framework.ConfigurationModel.UserSecrets/PathHelper.cs:
```diff
@@ -0,0 +1,57 @@
+            }
+
+            var root = Environment.GetEnvironmentVariable("APPDATA") ??         // On Windows it goes to %APPDATA%\Microsoft\UserSecrets\
+                        Environment.GetEnvironmentVariable("HOME");             // On Mac/Linux it goes to ~/Microsoft/UserSecrets/
```
comment is wrong. As they always are. Does it add that much value?

Praburaj replied:
Fixed.

davidfowl commented on test/SecretManager.Tests/UserSecretHelper.cs:
```diff
@@ -0,0 +1,47 @@
+
+namespace SecretManager.Tests
+{
+    public class UserSecretHelper
```
We need to mock out the file system. This kinda sucks.

glennc commented on src/SecretManager/CommandOutputLogger.cs:
```diff
@@ -0,0 +1,60 @@
+    /// <summary>
+    /// Logger to print formatted command output.
+    /// </summary>
+    public class CommandOutputLogger : ILogger
```
This feels more generic than this, is there nowhere else it can live?

davidfowl replied:
we had a discussion about that. It could love in logging but I dunno.

glennc replied:
Doesn't seem completely useless to be in logging. Otherwise the console options codebase?

Praburaj replied:
I'm going to file a work item to move this implementation somewhere else. Per the discussion we had for now this lives here.

glennc commented on src/Microsoft.Framework.ConfigurationModel.UserSecrets/ConfigurationExtensions.cs:
```diff
@@ -0,0 +1,33 @@
+        /// </summary>
+        /// <param name="configuration"></param>
+        /// <returns></returns>
+        public static IConfigurationSourceRoot AddUserSecrets([NotNull]this IConfigurationSourceRoot configuration)
```
Another overload where the appSecretId is passed in directly might be nice, seems more in-line with all our other apis.

Praburaj replied:
Done.

Praburaj commented:
Updated the PR addressing all feedback.

Praburaj commented:
@muratg @davidfowl @loudej @anurse @ChengTian - any more comments?

glennc commented:
I'm ok with this now, I want to get it merged and start using it a bit. Make more issues to improve it after that. :shipit:

Praburaj commented:
Thanks @glennc . Sure I can take up bugs / comments as it comes.

anurse commented:
:shipit:

muratg commented:
Looks good to me. :shipit:
